software/src/backends/hybrid/dev_backend.h

```diff
@@ -0,0 +1,413 @@
+   */
+  uint64_t ConvertVirtAddrToDevAddr(void* virt_addr) {
+    uint64_t phys_addr = virt_to_phys(virt_addr);
+
+    _enso_compiler_memory_barrier();
+    while (queue_to_backend_->Push(
+               {NotifType::kTranslAddr, (uint64_t)phys_addr, 0}) != 0) {
+    }
+
+    std::optional<PipeNotification> notification;
+
+    // Block until receive.
+    while (!(notification = queue_from_backend_->Pop())) {
+    }
+
+    assert(notification->type == NotifType::kTranslAddr);
+    assert(notification->data[0] == (uint64_t)phys_addr);
+
+    return notification->data[1];
+  }
+
+  /**
+   * @brief Retrieves the number of fallback queues currently in use.
+   * @return The number of fallback queues currently in use. On error, -1 is
+   *         returned and errno is set appropriately.
+   */
+  int GetNbFallbackQueues() {
+    struct PipeNotification pipe_notification;
+    pipe_notification.type = NotifType::kGetNbFallbackQueues;
+    while (queue_to_backend_->Push(pipe_notification) != 0) {
+    }
+
+    std::optional<PipeNotification> notification;
+
+    // Block until receive.
+    while (!(notification = queue_from_backend_->Pop())) {
+    }
+
+    assert(notification->type == NotifType::kGetNbFallbackQueues);
+    return notification->data[0];
+  }
+
+  /**
+   * @brief Sets the Round-Robin status.
+   *
+   * @param enable_rr If true, enable RR. Otherwise, disable RR.
+   *
+   * @return Return 0 on success. On error, -1 is returned and errno is set.
+   */
+  int SetRrStatus(bool enable_rr) {
+    struct PipeNotification pipe_notification;
+    pipe_notification.type = NotifType::kSetRrStatus;
+    pipe_notification.data[0] = (uint64_t)enable_rr;
+    while (queue_to_backend_->Push(pipe_notification) != 0) {
+    }
+    std::optional<PipeNotification> notification;
+
+    // Block until receive.
+    while (!(notification = queue_from_backend_->Pop())) {
+    }
+
+    assert(notification->type == notiftype::kSetRrStatus);
+    return notification->data[0];
+  }
+
+  /**
+   * @brief Gets the Round-Robin status.
+   *
+   * @return Return 1 if RR is enabled. Otherwise, return 0. On error, -1 is
+   *         returned and errno is set.
+   */
+  int GetRrStatus() {
+    struct PipeNotification pipe_notification;
+    pipe_notification.type = NotifType::kGetRrStatus;
+    while (queue_to_backend_->Push(pipe_notification) != 0) {
+    }
+    std::optional<PipeNotification> notification;
+
+    // Block until receive.
+    while (!(notification = queue_from_backend_->Pop())) {
+    }
+
+    assert(notification->type == notiftype::kGetRrStatus);
+    return notification->data[0];
+  }
+
+  /**
+   * @brief Allocates a notification buffer.
+   *
+   * @return Notification buffer ID. On error, -1 is returned and errno is set.
+   */
+  int AllocateNotifBuf(uint32_t application_id) {
+    struct PipeNotification pipe_notification;
+    pipe_notification.type = NotifType::kAllocateNotifBuf;
+    pipe_notification.data[0] = application_id;
```

Comment:
send application ID and TID for shinkansen to store associated with this notification buffer